src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -102,8 +106,288 @@ final class HashMap[K, +V] private[immutable] (private val rootNode: MapNode[K,
+
+                var rightDataIdx = 0
+                val rightPayloadArity = rightBm.payloadArity
+
+                var leftNodeIdx = 0
+                val leftNodeArity = leftBm.nodeArity
+
+                var rightNodeIdx = 0
+                val rightNodeArity = rightBm.nodeArity
+
+                val nextShift = shift + Node.BitPartitionSize
+
+                var idx = 0 // the conceptual, "un-compressed" index
+
+                var compressedDataIdx = 0
+
+                var compressedNodeIdx = 0
+
+                while (
+                  leftDataIdx < leftPayloadArity ||
+                    rightDataIdx < rightPayloadArity ||
+                    leftNodeIdx < leftNodeArity ||
+                    rightNodeIdx < rightNodeArity) {
+
+                  val bitpos = Node.bitposFrom(idx)
+
+                  if ((bitpos & leftNodeRightNode) != 0) {
+                    val newNode = concat(leftBm.getNode(leftNodeIdx), rightBm.getNode(rightNodeIdx), nextShift)
+                    result.content(newContentSize - compressedNodeIdx - 1) = newNode
+                    compressedNodeIdx += 1
+                    rightNodeIdx += 1
+                    leftNodeIdx += 1
+                    result.size += newNode.size
+
+                  } else if ((bitpos & leftDataRightNode) != 0) {
+                    val newNode = {
+                      val n = rightBm.getNode(rightNodeIdx)
+                      val leftKey = leftBm.getKey(leftDataIdx)
+                      val leftValue = leftBm.getValue(leftDataIdx)
+                      val leftOriginalHash = leftBm.getHash(leftDataIdx)
+                      val leftImproved = improve(leftOriginalHash)
+
+                      if (n.containsKey(leftKey, leftOriginalHash, leftImproved, nextShift)) {
+                        newHash -= improve(leftBm.getHash(leftDataIdx))
+                        n
+                      } else {
+                        n.updated(leftKey, leftValue, leftOriginalHash, leftImproved, nextShift)
+                      }
+                    }
+
+                    result.content(newContentSize - compressedNodeIdx - 1) = newNode
+                    compressedNodeIdx += 1
+                    rightNodeIdx += 1
+                    leftDataIdx += 1
+                    result.size += newNode.size
+                  }
+                  else if ((bitpos & leftNodeRightData) != 0) {
+                    val newNode = {
+                      val n = leftBm.getNode(leftNodeIdx)
+                      val rightKey = rightBm.getKey(rightDataIdx)
+                      val rightValue = rightBm.getValue(rightDataIdx)
+                      val rightOriginalHash = rightBm.getHash(rightDataIdx)
+                      val rightImproved = improve(rightOriginalHash)
+
+                      if (n.containsKey(rightKey, rightOriginalHash, rightImproved, nextShift)) {
+                        newHash -= improve(rightBm.getHash(rightDataIdx))
```

Comment:
Could we base this hash adjustment on `newNode.size == n.size` instead?

Reply:
Great idea!